webapp/channels/src/components/channel_notifications_modal/components/notification_section.tsx

```diff
@@ -1,112 +1,120 @@
+
 import CollapseView from './collapse_view';
 import ExpandView from './expand_view';
 
-export default class NotificationSection extends React.PureComponent {
-    static propTypes = {
+type SelectedOption = {
+    label: string;
+    value: string;
+};
 
-        /**
-         * Notification section
-         */
-        section: PropTypes.string.isRequired,
+export type Props = {
 
-        /**
-         * Expand if true, else collapse the section
-         */
-        expand: PropTypes.bool.isRequired,
+    /**
+     * Notification section
+     */
+    section: string;
 
-        /**
-         * Member's desktop notification level
-         */
-        memberNotificationLevel: PropTypes.string.isRequired,
+    /**
+     * Expand if true, else collapse the section
+     */
+    expand: boolean;
 
-        memberDesktopSound: PropTypes.string,
+    /**
+     * Member's desktop notification level
+     */
+    memberNotificationLevel: string;
 
-        memberDesktopNotificationSound: PropTypes.string,
+    memberDesktopSound?: string;
 
-        /**
-         * Member's desktop_threads notification level
-         */
-        memberThreadsNotificationLevel: PropTypes.string,
+    memberDesktopNotificationSound?: string;
 
-        /**
-         * Ignore channel-wide mentions @channel, @here and @all
-         */
-        ignoreChannelMentions: PropTypes.string,
+    /**
+     * Member's desktop_threads notification level
+     */
+    memberThreadsNotificationLevel?: string;
 
-        /**
-         * Auto-follow all new threads in this channel
-         */
-        channelAutoFollowThreads: PropTypes.string,
+    /**
+     * Ignore channel-wide mentions @channel, @here and @all
+     */
+    ignoreChannelMentions?: string;
 
-        /**
-         * User's global notification level
-         */
-        globalNotificationLevel: PropTypes.string,
+    /**
+     * Auto-follow all new threads in this channel
+     */
+    channelAutoFollowThreads?: string;
 
-        /**
-         * User's global notification sound
-         */
-        globalNotificationSound: PropTypes.string,
+    /**
+     * User's global notification level
+     */
+    globalNotificationLevel?: string;
 
-        /**
-         * onChange handles update of desktop notification level
-         */
-        onChange: PropTypes.func.isRequired,
+    /**
+     * User's global notification sound
+     */
+    globalNotificationSound?: ChannelNotifyProps['desktop_notification_sound'];
 
-        /**
-         * onChangeThreads handles update of desktop_threads notification level
-         */
-        onChangeThreads: PropTypes.func,
+    /**
+     * onChange handles update of desktop notification level
+     */
+    onChange: (value: string | any) => void;
 
-        onChangeDesktopSound: PropTypes.func,
+    /**
+     * onChangeThreads handles update of desktop_threads notification level
+     */
+    onChangeThreads?: (value: string | any) => void;
```

Comment:
These all have fleshed out types here: https://github.com/ivalkshfoeif/mattermost/blob/e7721bc95ffa4cd9aedefce938e9dd0831c884a6/webapp/channels/src/components/channel_notifications_modal/components/expand_view.tsx#L32

Reply:
Hi, for serverError, it is string | null in expand_view, do I need to update it as well?
for onChangeThreads, if I replace it with `(e: ChangeEvent<HTMLInputElement>) => void`, it will conflict with the following code which takes a string as input.

```typescript
    handleOnChangeThreads = (e: ChangeEvent<HTMLInputElement>) => {
        const value = e.target.checked ? NotificationLevels.ALL : NotificationLevels.MENTION;

        this.props.onChangeThreads?.(value);
    };
```

Reply:
Ahh okay I missed that. We do know that we're passing a `string` though from an input, so can we remove the `any`?

Reply:
The input type is `ChannelNotifyProps['mark_unread']` in [channel_notifications_modal](https://github.com/mattermost/mattermost/blob/master/webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx) for onChange input, if I remove `any` from input option, it can't automatically consider it as string.

Reply:
Is 'mark_unread' just a constant? If so we can use that in place of `any`:
`string | 'mark_unread'`

Reply:
It is actually ` 'all' | 'mention' `, can I use replace it with `string | 'all' | 'mention'`?

Reply:
The usage of onChange function includes:
```typescript
onChange={this.handleUpdateMarkUnreadLevel}

(property) ChannelNotificationsModal.handleUpdateMarkUnreadLevel: (markUnreadNotifyLevel: "all" | "mention") => void
```
```typescript
 onChange={this.handleUpdateDesktopNotifyLevel}

(property) ChannelNotificationsModal.handleUpdateDesktopNotifyLevel: (desktopNotifyLevel: "all" | "mention" | "default" | "none") => void

```

```typescript
onChange={this.handleUpdateMarkUnreadLevel}

(property) ChannelNotificationsModal.handleUpdateMarkUnreadLevel: (markUnreadNotifyLevel: "all" | "mention") => void

```

```typescript
onChange={this.handleUpdatePushNotificationLevel}

(property) ChannelNotificationsModal.handleUpdatePushNotificationLevel: (pushNotifyLevel: "all" | "mention" | "default" | "none") => void

```

Reply:
Can I replace them with `ChannelNotifyProps[<some constants>]`?

I find using `|` like
``` typescript
(value: ChannelNotifyProps['mark_unread'] | ChannelNotifyProps['ignore_channel_mentions']) => void;
```
will throw errors: `Type '"all" | "mention" | "default" | "off" | "on"' is not assignable to type '"all" | "mention" | "default" | "none"'.` 


but `&` can solve it like 
``` typescript
(value: ChannelNotifyProps['mark_unread'] & ChannelNotifyProps['ignore_channel_mentions']) => void;
```

Can I choose this way to fix the issue?